nixos/modules/config/stub-ld.nix

```diff
@@ -0,0 +1,82 @@
+{ config, lib, pkgs, ... }:
+
+let
+  inherit (lib) last splitString mkOption types mdDoc mkIf optionals;
+
+  message = ''
+    You seem to be trying to run a dynamically linked executable that
+    wasn't made for NixOS. This is possible, but not without extra steps,
+    because NixOS, by design, does not have a global library path.
+
+    There are several solutions available:
+
+    1. Use the version packaged in nixpkgs, if there is one. You can
+       search available packages at https://search.nixos.org/packages
+
+    2. Write a nix expression for the program to package it
+       yourself. There are multiple approaches to this:
+
+       a. Build from source. This is ideal. For an introduction to
+          this, see https://nix.dev/tutorials/packaging-existing-software
+
+       b. Fix the program's ELF metadata to hardcode paths to libraries
+          using autoPatchelfHook. Use this if building from source isn't
+          feasible.
+
+       c. Wrap the program to run in a normal-seeming container using
+          buildFHSEnv. This is undesirable, but sometimes necessary, for
+          example, if the package downloads and runs further executables.
+
+    3. Create a library path that only applies to unpackaged programs
+       using nix-ld. Add this to your configuration.nix:
+
+       programs.nix-ld.enable = true;
+       programs.nix-ld.libraries = with pkgs; [
+         # Add any missing dependencies for unpackaged programs here,
+         # NOT in environment.systemPackages
+       ];
+
+    4. Use the FHS-like container used by the steam package with
+       steam-run <command>. This is quick and easy, and usually works,
+       but isn't a particularly good long-term solution.
+  '';
+
+  rule-for = pkgs: let
+    stub-ld = pkgs.pkgsStatic.runCommandCC "stub-ld" {
+      nativeBuildInputs = [ pkgs.unixtools.xxd ];
+      inherit message;
+    } ''
+      printf "%s" "$message" | xxd -i -n message >main.c
+      cat <<EOF >>main.c
+      #include <stdio.h>
+      int main() {
+        fwrite(message, sizeof(unsigned char), message_len, stderr);
```

Comment:
Nice to have, but not a blocker: Can we somehow print what binary was executed here?
Often this program might be not started directly by the user but by other programs i.e. npm.

Reply:
I think it's also just fine to print the executable name?

Reply:
Yeah, that's what I ended up doing. The arguments aren't necessary.